orderer/common/multichannel/registrar.go

```diff
@@ -801,3 +801,60 @@ func (r *Registrar) loadJoinBlocks() map[string]*cb.Block {
 
 	return channelToBlockMap
 }
+
+func (r *Registrar) removeSystemChannel() error {
+	systemChannelID := r.systemChannelID
+	consensusType := r.systemChannel.ConsensusType()
+	if consensusType != "etcdraft" {
+		return errors.Errorf("cannot remove %s system channel: %s", consensusType, systemChannelID)
+	}
+
+	// halt the inactive chain registry
+	consenter := r.consenters["etcdraft"].(consensus.ClusterConsenter)
+	consenter.StopInactiveChainRegistry()
+
+	// halt all application channels
+	for channel, cs := range r.chains {
+		if channel == systemChannelID {
+			// system channel is halted after halting all app channels
+			delete(r.chains, channel)
+			continue
+		}
+		cs.Halt()
+
+		rl, err := r.ledgerFactory.GetOrCreate(channel)
+		if err != nil {
+			return errors.Wrapf(err, "could not retrieve ledger for channel: %s", channel)
+		}
+		configBlock := ConfigBlockOrPanic(rl)
+		isChannelMember, err := consenter.IsChannelMember(configBlock)
+		if err != nil {
+			return errors.Wrapf(err, "failed to determine channel membership for channel: %s", channel)
+		}
+		if !isChannelMember {
+			logger.Debugf("Not a member of channel %s, removing it", channel)
+			err := r.removeMember(channel, cs)
+			if err != nil {
+				return errors.Wrapf(err, "failed to remove channel: %s", channel)
+			}
+		}
+	}
+
+	// halt the system channel
+	r.systemChannel.Halt()
+
+	err := r.ledgerFactory.Remove(systemChannelID)
+	if err != nil {
+		return errors.Errorf("error removing ledger for system channel %s", r.systemChannelID)
+	}
+
+	r.systemChannel = nil
+	r.systemChannelID = ""
+
+	// reintialize the registrar to recreate all the channels
+	r.Initialize(r.consenters)
```

Comment:
In light of the lock comment, this will deadlock as the locks are not re-entrant... 

I think we need to separate the `r.Initialize(consenters)` into a `init()` and a `startAll()` that can be called from here within the lock.  That is, Initialize will call init() w/o a lock, then startAll() with a lock. Here we would call both with a lock.

Reply:
I wasn't sure if I should split things out so I left it as is before and that resulted in leaving out the locks elsewhere. Thanks for suggesting this. Done.